onnxruntime/core/providers/openvino/backends/basic_backend.cc

```diff
@@ -64,20 +64,14 @@ BasicBackend::BasicBackend(const ONNX_NAMESPACE::ModelProto& model_proto,
   }
   LOGS_DEFAULT(INFO) << log_tag << "Loaded model to the plugin";
 
-  // Create infer request
-  try {
-    infer_request_ = exe_network.CreateInferRequestPtr();
-  } catch (InferenceEngine::details::InferenceEngineException e) {
-    ORT_THROW(log_tag + " Exception while creating InferRequest object: " + e.what());
-  } catch (...) {
-    ORT_THROW(log_tag + "Exception while creating InferRequest object");
-  }
-  LOGS_DEFAULT(INFO) << log_tag << "Infer request created";
+  //Initializing the infer_requests_ pool with 8 infer_request's (creating infer_request)
+  size_t nireq = 8;
```

Comment:
should this be made configurable?

Reply:
> should this be made configurable?

@jywu-msft George, with nireq = 8, a pool of 8 usable InferRequests will be made available which may be picked up by any of the threads running in parallel. 

Since, we don't know what might be the optimal thread number for target setting for a particular device and which would be only known by some study and benchmarking numbers for that specific hardware device, we decided to go with max of 8 because this is what the maximum value that even the OpenVINO  picks as an optimal number on an CPU during async inference.

so, In the email thread, we decided to set **thread number 3** as the target setting for an MYRIADX after our study. But we might require to increase this number to get a optimal number for a CPU device with many cores and other devices as well.

so, we wanted to have a max of 8 infer_requests_ available in the pool and let the user/developer decide the target setting based on their device they are trying to run based on their study and performance numbers.

Reply:
we shouldn't limit this setting for Myriadx right? basic_backend also work for other device.

---
In reply to: [510388374](https://github.com/microsoft/onnxruntime/pull/5568#discussion_r510388374) [](ancestors = 510388374)

Reply:
> we shouldn't limit this setting for Myriadx right? basic_backend also work for other device.
> 
> In reply to: [510388374](https://github.com/microsoft/onnxruntime/pull/5568#discussion_r510388374) [](ancestors = 510388374)

@HectorSVC No we are not limiting the setting to any particular device here. It's just that we will have max of 8 free infer request's available in the pool. you can run any device using the basic_backend.

what I was trying to tell is that, so prior to this PR, we use to have support for only 1 Thread. Now, with Multi Threading enabled, we can run multiple threads in parallel. Now, the user can decide, what should be the optimal number of threads for his application and the device he wants to run the application on.

This applies to all the hardware devices we support via basic_backend today.

**Just for Example:**  Say, a user wants to run an application on iGPU using basic_backend and he wants to make use of multi_threading. Now, the user performs some experiments and realizes that with 5 threads he is getting the least total inference time for 1000 request's. so, now he will go on to use **-c 5** for his application.

Say, for the same device iGPU, the same user for running another application decides that he wants to use 4 threads for this use case to play the tradeoff between throughput and latency. Now, he will use **-c 4** for his application.

For the very same reason, we don't want to fix an optimal number for each device. so, we decided to have max of 8 infer request's available in the pool and let the user decide while running his application the thread setting **(-c value)** he want's to go with.

Reply:
I think our question is whether max of 8 is sufficient or whether it needs to be made configurable.

Reply:
@jywu-msft sure george, I'll make this value configurable and push my changes

Reply:
> @jywu-msft sure george, I'll make this value configurable and push my changes

@jywu-msft  Added changes to make nireq value configurable during runtime. If this value is not specified at runtime, default value of 8 would be used during build time.